proposals/0082-relax-transaction-constraints.md

```diff
@@ -0,0 +1,173 @@
+---
+simd: '0082'
+title: Relax Transaction Constraints
+authors:
+  - Andrew Fitzgerald (Solana Labs)
+category: Standard
+type: Core
+status: Draft
+created: 2023-10-30
+feature:
+---
+
+## Summary
+
+This proposal aims to relax some of the constraints on which individual
+transactions can be included in a valid block.
+The proposal does not relax the constraints required for a transaction to be
+executed.
+
+## Motivation
+
+The current protocol places many constraints on the structure and contents
+of blocks; if any constraints are broken, block-validators will mark the block
+as invalid.
+Many of these constraints are necessary, but some of them are not, and lead to
+additional complexity in the protocol.
+This proposal aims to relax some of the constraints at the individual
+transaction level, in order to simplify the protocol, and give more flexibility
+to block-producer and block-validator implementations.
+
+More specifically, this proposal aims to relax the constraints which require
+account state in order to determine the validity of a transaction's inclusion.
+The reason these constraints are targetted specifically, is that reliance on
+account state necessarily requires synchronous execution within the protocol.
+This proposal on its' own, will not enable asynchronous execution, but it will
+remove one of the barriers to asynchronous execution.
+
```

Comment:
A bit confused by "aim" in Summary section and Motivations here. As "aim" seems to yield benefit to block producer ("...transaction can be included..."), which is not part of protocol; while Motivation talks about modifying protocol. 

Is it correct yo say this proposal has two parts: 
1. remove listed constraints/requirement of transaction for block producer to include them into block; this part solely in Producer implementation, does not modify protocol;
2. instead of invalidate entire block at earlist failure of transaction due to named constraints, validator only _not_ execute bad transactions but will continue validate the block. (bad transactions included in ledger). This part of protocol altering.

or maybe I am thinking on wrong track?

Reply:
I think only your 2nd point is included in the proposal. Block-production changse are enabled by the changes in the proposal, but not required beyond potential changes to the fee-colection.

> instead of invalidate entire block at earlist failure of transaction due to named constraints, validator only not execute bad transactions but will continue validate the block. (bad transactions included in ledger). 

yes. Those bad transactions may end up having limited effect if we are still able to collect fees, or if we decide to make the block-producer cover those fees. They will not be executed though.

----

Proposal does not suggest or require the block-producer to make any changes, but allows for changes. For several of the constraints, the producer is still incentivized to perform the checks (fee-payer, nonce) because otherwise they will not receive rewards for those transactions. There's freedom in how that is done, and the block-producer could potentially do this in a non-strict fashion.

I'll cover this in the updated prop, but essentially block-validation and consensus really consists of 2 primary questions:

1. is this block valid?
2. what is the result of the block?

The changes here are intending to make answering this first question, "is this block valid?", a simpler process that does not require executing the block. This proposal separates these questions logically, but w/ current voting we still require the 2nd question, "what is the result of the block?", be answered synchronously.